### ISSUE TITLE
Support for getting terminal width & height

### ISSUE THREAD

**Issue by derekdreery** [Stebalien/term#50]
Getting the terminal width and height is something you cannot do presently with the `Terminal` trait. It prevents implementation of something like [termbox](https://github.com/nsf/termbox) on top of `Terminal` trait.

The linux implementation looks something like (from termbox)

```c
	struct winsize sz;
	memset(&sz, 0, sizeof(sz));

	ioctl(inout, TIOCGWINSZ, &sz);

	termw = sz.ws_col;
	termh = sz.ws_row;
```

whilst in windows the relevant function is `kernel32::GetConsoleScreenBufferInfo`

**Comment by derekdreery** [comment]
I've had a look at implementing this - the main problem as I see it is that currently the unix implementation does everything with terminfo, therefore not needing to make any system calls directly. The best way to get the console dimensions is to issue an ioctl to populate the `winsize` struct, but that may not be possible with serial consoles for example.

According to https://unix.stackexchange.com/questions/93173/how-does-less-know-the-terminal-resolution the program `less` takes the approach of using ioctl if available, and then using some tricks to try and figure out the width/height (e.g. environment variables).

So questions:
 1. Do you want/mind linking to libc on linux?
 2. Do you therefore think that the size function should return an Option, with none if it wasn't possible to calculate the size?

**Comment by derekdreery** [comment]
You can see my messings about at https://github.com/derekdreery/term :master

**Comment by Stebalien** [comment]
1. I'm fine with that.
2. When #49 gets merged, we can use a `NotSupported` error variant.

As for getting this working on unix, given that #49 will be a breaking change anyways I plan on reworking the internals quite a bit (after it lands) so this will probably get much easier.

**Comment by derekdreery** [comment]
http://docs.freebsd.org/44doc/usd/13.viref/paper.html#section5f seems like a good strategy?

**Comment by Stebalien** [comment]
@derekdreery Yes, that looks good. However, I'm probably going to reorder those a bit (use ioctl first, then the environment, then the terminfo database.

**Comment by xSke** [comment]
Would this be possible in Windows too?

**Comment by Stebalien** [comment]
Yes but I won't be able to do that myself (I don't use windows).

**Comment by derekdreery** [comment]
I put the windows function on the top of this thread - I can implement it - [waiting on some constants in libc at the moment](https://github.com/rust-lang-nursery/libc/pull/123)

**Comment by derekdreery** [comment]
... a long time later

I've submitted PR #77 to implement this.
